### ISSUE TITLE
Xiaomi Mijia 1080P FHD Smart IP Camera - Support

### ISSUE THREAD

**Issue by EliasKotlyar** [EliasKotlyar/Xiaomi-Dafang-Hacks#624]
This issue is for the owners of Xiaomi Mijia:
https://www.gearbest.com/ip-cameras/pp_009981882806.html?wid=1433363

Here is a bootlog(thanks @gabest11)
https://github.com/gabest11/Xiaomi-Mijia-1080p-2018/blob/master/bootlog.md

It has probably the following specs(according to bootlog):
T20 with 64Mb
New Sensor named "ps5250" at address : 0x48
rtl8189fs Wifi

**Comment by gabest11** [comment]
Okay, I should receive the flash programmer in a few weeks. I will continue then. Even if there is no easy way to hack the firmware, the dafang scripts and binaries should be re-usable.

**Comment by EliasKotlyar** [comment]
@gabest11: 
About the Firmware: You will need a special bootloader for achieving this - you will have to compile it. You wont be able to use your normal bootloader from nand.

I suppose that you can flash a image by providing a tf_recovery.img file. The format will be (hopefully) the same as on the normal dafang. I would wait for the programmer, read the firmware out and then try to flash it using this file. I say, that a CFW from a Xiaofang 1S should boot without any modifications.

So if you have 2 of them, you can try to flash one and check. If you have only one, wait for the programmer to arrive, because you would not be able to recover without the "original" files.

**Comment by gabest11** [comment]
Yes, I definitely want to save the original fw first. For the unique data that identifies it to xiaomi's servers.

**Comment by gabest11** [comment]
I am uncertain if a tf_recovery.img can be constructed. The old mijia 1080p used the standard firmware image (output of mkimage), but to feed it from the sd card, it had to be signed with a 16 byte hash at the end of the file. It is not a simple md5, or not what the standard md5 algorithm produces.

**Comment by Blaize2345** [comment]
I have this white version and can confirm that I tried countless TF_recovery.img files on it and different methods of long press reset, etc to have it restore to original firmware. Nothing.

The device was bricked during an official update via the MI HOME app.

If you want me to test any created firmware that can be placed on an SD Card for you guys, let me know. Thumbs up to the work you're doing on this white version with new firmware. The back of the camera is sealed and all solid white.

Model is: SXJ02ZM
Current state of hardware: Slow Yellow Blink. No other response from camera.

**Comment by EliasKotlyar** [comment]
@Blaize2345 : Can you try to solder the serial connection to the camera? It may be a bootloader-brick, which cannot be recovered using microsd....

**Comment by gabest11** [comment]
The recovery images floating on the internet are for the old version, a different architecture. It would not work even if you somehow managed to load it into its flash chip. I am not aware of anything for the new model. Maybe there is one on some Chinese forums, but they are pretty hard to get, most of the time they just PM each other the files.

**Comment by Blaize2345** [comment]
That's indeed correct. I hoping to find (or wait for) an official extract of the official firmware for this WHITE model SXJ02ZM first before attempting to open it up.... Surely a way to restore it from a bad firmware flash must exist.

Xiaomi should revisit and retest their flashing mechanism from the Mi Home App. This bricking syndrome is way too common. Though Google Translate I realised that on xiaomi.cn there is great demand for an official firmware to be 'brushed' so they can restore these devices to normality. (brushed is 'flashed' according to Google Translate from what I could understand).

Hoping for the magic official firmware recovery file (or method) for the SXJ02ZM then.

**Comment by Blaize2345** [comment]
UPDATE - As suspected, there is a SPECIFIC tf_recovery.img for the ALL-WHITE model SXJ02ZM Xiaomi Mi 1080P Smart security cam.

And I finally found found it! Attached.

Surprisingly, the file is referred to in the internal MI HOME APP help system. But if you have a bricked cam you can't see this specific help section. 

Attached in this post! Keep safe!

[tf_recovery for SXJ02ZM All White Xiaomi 1080P smart cam.zip](https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/files/2320611/tf_recovery.for.SXJ02ZM.All.White.Xiaomi.1080P.smart.cam.zip)

**Comment by gabest11** [comment]
All the same software as the old one, but for a different hardware. They kept the backdoor to boot from the sd card, but the leaked key to sign the md5 hash of the init script was replaced long ago, obviously. We can always take it apart and flash a modded /etc/init.d/rcS with the right tools, but it is pretty annoying to do.

**Comment by gabest11** [comment]
GM had config files for day and night mode, it is not the case this time. The image is a over-exposed with sunlight and the night vision is not sharp at all.

**Comment by willthrom** [comment]
Hi,

I saw your are having problems with the V3.

I was to say if you took a look to the previous code I did
https://github.com/Filipowicz251/mijia-1080P-hacks

But at the end I checked myself downloading the recovery.

And yes, it seems so far the only way is to create a new factoryImage but that would be tricky as they are normally signed.....

And yes as well, in this version the signature for the SDCARD access is the public one (last time in the 99 version or recovery the signature was the private). They fixed later but it was still possible to bypass it.

Has anyone try to load the firmware from the other camera? you might be lucky.
If that is the case, even if half of the camera doesn't work. from that firmware you can upgrade to the proper one after altering the path of the KEYS/Binaries, etc....

**Comment by cadumantovane** [comment]
Sorry, but I'm a layman and I need help.
I have 4 Xiaomi Mijia 1080p IP Camera logged into Mi Home in OTHER region (I'm from Brazil). Now I bought a All White (Model: SXJ02ZM) and I can't log in if it is not in Mainland China. The current firmware is 3.4.5_0041. Is there any way to make it log in to the other region along with the others?

**Comment by willthrom** [comment]
that is what I suggested to do... if someone can try:

https://github.com/Filipowicz251/mijia-1080P-hacks/issues/55#issuecomment-421043218

**Comment by jaaperror** [comment]
Hi everyone,

I am very interested in a hacked firmware for the ALL WHITE version of this cam Model: SXJ02ZM). I am not a developer though but is there anything I can do? Someone I can buy a coffee? Hoping for the always desired RTSP possibility :)

**Comment by Anfero** [comment]
@Blaize2345, could you please tell me how did you too recover it from the brick?
I have downloaded the recovery .img in the link and put it in a new micro SD, then switched off the camera, put the micro SD and turn on, but nothing has changed. The orange light switches on and off every three seconds +/-

**Comment by Anfero** [comment]
The camera is the same model as yours, I'll try formatting it in case it
came in any other format type.
Thanks for your fast reply and taking your time replying

**Comment by Anfero** [comment]
I just did it and it's still the same.
I even downloaded a program to format the card in fat 32 but when it
finished it said that it is exFat instead of Fat32, maybe the issue is
there?
Also my micro SD card is a 64Gb capacity, may this be an issue?
I can't think anything else...
Thanks for your time.

**Comment by Blaize2345** [comment]
Yes that's the issue. You can't fat32 format a 64gb card! 

Find a 2gb or smaller sdcard and format that to fat32. Copy file, flash and it will work.

**Comment by Anfero** [comment]
Thanks, I'll try to find it

**Comment by gbarral** [comment]
Hi, i try the tf_recovery.img given by Blaize2345 whith the hack https://github.com/Filipowicz251/mijia-1080P-hacks. The tf_recovery seems to work because the camera downgrade (3.4.4_0039) but the Tools is not installed. Impossible to connect using SSH. I can update 3.4.5_0046 whith mi-home but impossible to activate RSTP.

If anybody have idea :-)

**Comment by mmartain** [comment]
Recovery image worked for me! Formatted my 64GB card with 1GB FAT32 partition, loaded recovery and had to pull plug about 5 minutes. Afterwards rebooted without sd card and camera started talking. Setup like normal again.

**Comment by axlerose** [comment]
Hi there is any news? I've two of this camera and i can make some test if necessary

**Comment by EliasKotlyar** [comment]
@axlerose : Unfortunately no news yet. If someone can hook up a serial connection to his camera and then can dump the whole flash, i can try to make a firmware

**Comment by axlerose** [comment]
I've see that @gabest11  have just open the camera https://github.com/gabest11/Xiaomi-Mijia-1080p-2018

**Comment by EliasKotlyar** [comment]
@axlerose Many thanks for the pictures. Can you please desolder the chip before trying to read it? I had this problem already with my Dafang. I was not able to read the chip when it was soldered down to the board.

**Comment by axlerose** [comment]
@EliasKotlyar i'm not able o desolder a soc i've not this Knowledge but i Think that @gabest11 can

**Comment by gabest11** [comment]
I think my reader does not support this chip, the identifier is not on the list. But getting the fw was never the problem, it can be downloaded from the android app's help section and unpacked/repacked. There are some extra bytes, hash or signing at the end, which I cannot reproduce.

**Comment by EliasKotlyar** [comment]
It would really help if you would read the whole firmware. The firmware update package mostly does not contain all the flash parts. Especially the bootloader seems to be the weird part. If we would analyze that, it would help us to understand how its behaving  and maybe figure out a solution. I can try to provide you a open-source bootloader, as soon as you figure out how to read/write it to the nand-flash.

**Comment by anmaped** [comment]
@EliasKotlyar You can check the dump of the flash [here](https://github.com/anmaped/openfang/blob/master/doc/sxj02ZM/dump.tar.xz).

@gabest11 It is possible to open the camera without destroing it. You just need two credit cards; check it out [here](https://github.com/anmaped/openfang/blob/master/doc/sxj02ZM/pictures/20181214_071923.jpg) !

**Comment by EliasKotlyar** [comment]
@anmaped : Many thanks! It seems that it has the same layout as the normal one, except that the the whole firmware is somehow "obfuscated". I tried to flash it onto my Dafang, but it was not successfull. Can you try to flash a custom bootloader and boot into custom-firmware? It will help analyzing the firmware a lot.  Just try to flash the "dafang_64mb_v1.bin" as a first try.

**Comment by jeancleber** [comment]
@anmaped do you know where are the serial pins in the mainboard of this camera? I guess is near the flash ic but I don’t know the correct sequence. Thanks!

**Comment by axlerose** [comment]
@anmaped have you any news? i've try to open the camera but i didn't find the tx-rx

**Comment by jeancleber** [comment]
OK. Using the photo as provided by @gabest11 , here are the pinouts for the serial interface.
Tried to stop the U-boot = no success.
Tried to SSH via ports 22 and 21 = no success.
Tried to load the Xiaofang 1S bootloader from SD card = no success.
Through serial port, I can only monitor what is going on with the device...

@anmaped, you wrote in other topic (https://github.com/Filipowicz251/mijia-1080P-hacks/issues/55#issuecomment-447280626) that you successfully ran openfang in a white-back Mijia 1080p camera... how did you do that? How can you get SSH in this device?
 
Thanks.
![mijia 1080p all white pinout](https://user-images.githubusercontent.com/17370317/50380185-77dd9b00-0646-11e9-873a-00a154cc5eb1.jpg)

**Comment by CansuErgin** [comment]
@Blaize2345 hello, I tried with the file you gave me. After the "click", the orange light of the camera flashes. After the second "click" then continued to burn again fixed. I switched off the device and removed the 16gb sd card and then switched on the device again, but the orange light continued to be lit. No sound and not reset. I ran it with the reset button but I didn't get a result.

do I have something to jump or jump?

thank you so much.

**Comment by CansuErgin** [comment]
1gb microsd card partitioned. I threw the recovery file in it. I installed the sd card and switched on the device. The orange light starts blinking. flashed for about a minute, and then click sound came on again and continued to burn. I've waited 5 minutes. then I unplugged the device and removed the sd card. I switched on the device again, but the orange light is still on. am I doing something wrong?

**Comment by eflukx** [comment]
> OK. Using the photo as provided by @gabest11 , here are the pinouts for the serial interface.
> Tried to stop the U-boot = no success.
> Tried to SSH via ports 22 and 21 = no success.
> Tried to load the Xiaofang 1S bootloader from SD card = no success.
> Through serial port, I can only monitor what is going on with the device...
> 
> @anmaped, you wrote in other topic ([Filipowicz251/mijia-1080P-hacks#55 (comment)](https://github.com/Filipowicz251/mijia-1080P-hacks/issues/55#issuecomment-447280626)) that you successfully ran openfang in a white-back Mijia 1080p camera... how did you do that? How can you get SSH in this device?
> 
> Thanks.
> ![mijia 1080p all white pinout](https://user-images.githubusercontent.com/17370317/50380185-77dd9b00-0646-11e9-873a-00a154cc5eb1.jpg)

RX and TX seem to be reversed.

**Comment by effendieffendi** [comment]
https://www.winbond.com/resource-files/w25q128jv%20revf%2003272018%20plus.pdf

**Comment by phongtygen** [comment]
Hi,

I need to download firmware (tf_recovery) for this model of Mi Home Security Camera 360.

Model: MIXJ04CM

If someone knows where to download it, please share the link. Thanks a lot!

**Comment by anmaped** [comment]
For who wants to contribute here it is a map of the known pins in dafang and mijia cameras.

@EliasKotlyar Do you remember more on other cameras?

https://docs.google.com/spreadsheets/d/1NTUvBG2uCqyoVK8Nu5Ru6x2F3q2NCy82xTfKQ9fuavU

**Comment by anmaped** [comment]
> 
> 
> @anmaped : Many thanks! It seems that it has the same layout as the normal one, except that the the whole firmware is somehow "obfuscated". I tried to flash it onto my Dafang, but it was not successfull. Can you try to flash a custom bootloader and boot into custom-firmware? It will help analyzing the firmware a lot. Just try to flash the "dafang_64mb_v1.bin" as a first try.

Just for curiosity the layout is different.
```
("boot", 0x40000),
("kernel", 0x280000),
("rootfs", 0x6F0000),
("data", 0x630000),
("config", 0x10000),
```

**Comment by hoxsiew** [comment]
Has there been any work on the "All White" camera?  I picked one up a few days ago and it apparently isn't compatible with any of the hacks I've seen.

I was able to get wires soldered to the TX, RX and GND pins and get a serial console, but u-boot has no interrupt pause and it immediately boots into the system.  I was able to glitch it by shorting the CLK pin of the eeprom as soon as I saw the u-boot init.  this prevented the kernel image from loading and gave me a u-boot prompt.  From there I was able to modify the kernel command line to use '/bin/ash' as init and it booted and dropped me into a shell.  From there I was able to launch linuxrc as a sub-process and still keep a root shell.  After that, I copied all the partitions to an SD card for further examination.

I have full control now of the device, but I'm not sure how to proceed.  The firmware hack I've seen don't seem to match the partition layout or kernel versions on my device so I'm hesitant to try and flash anything to the device.

**Comment by pcmester** [comment]
@hoxisew
There is the hack for the SXJ02ZM
https://github.com/Filipowicz251/mijia-1080P-hacks/issues/55

**Comment by hoxsiew** [comment]
Thanks for the link.  Do you think I can make that work without having to manually flash the eeprom?  I think I should be able to flash the images from existing u-boot or from the root shell and not have to molest the circuit board.

**Comment by pcmester** [comment]
I hope, because I would like to do that without de and re soldering the chip

**Comment by hoxsiew** [comment]
Well, it looks like I get to try.  Using `dd' to copy the u-boot image resulted in a soft-brick.  I get a few lines from the terminal then it hangs:
`U-Boot SPL 2013.07 (Nov 21 2018 - 00:42:00)
pll_init:365
l2cache_clk = 375000000
pll_cfg.pdiv = 8, pll_cfg.h2div = 4, pll_cfg.h0div = 4, pll_cfg.cdiv = 1, pll_cfg.l2div = 3
nf=36 nr = 1 od0 = 1 od1 = 1
cppcr is 02404900
CPM_CPAPCR 0470890d
nf=42 nr = 1 od0 = 1 od1 = 1
cppcr is 02a04900
CPM_CPMPCR 07d0c90d
nf=50 nr = 1 od0 = 1 od1 = 1
cppcr is 03204900
CPM_CPVPCR 0320490d
cppcr 0x9a794410
apll_freq 860160000 
mpll_freq 1000000000 
vpll_freq = 1200000000
ddr sel mpll, cpu sel apll
ddrfreq 500000000
cclk  860160000
l2clk 286720000
h0clk 250000000
h2clk 250000000
pclk  125000000
DDRC_DLP:0000f003
`
I don't have a CH341A but I do have a handful of RPi zeros.  I should be able to use the SPI bus to directly connect to the chip.  I'll report back.

**Comment by hoxsiew** [comment]
I built a RPi Zero SPI eeprom programmer using flashrom as the code and it works.  I was able to lift pin 5 from the chip and the programmer worked fine.  While the chip pin-5 was unsoldered, I tried several different u-boot images from releases and from development branches, and by holding the contact for pin-5 to the board, test them.  None of them went any further than the dump above.  I think there may be something very different in the architecture of this SoC, or I'm just not as competent as I thought.

I've replaced the original firmware and added a custom init to the rootfs so that I can get in without doing tricks with u-boot.  I basically just call /bin/ash in init and from there I can call the original script, or do whatever.  I'll try to see where I can go from here.

**Comment by goermezer** [comment]
@hoxsiew, I just have bricked my Xiaofang 1s and would offer me as a tester. I would like to flash either the original boot loader or uboot. I have RPI Zeros as well as RPI 2/3s available. I think, all of them will work, because all RPIs have an SPI interface, which first must be activated. 

I ordered a ch341 spi nand programmer and studied https://github.com/Dafang-Hacks/spiflasher. Meanwhile I want to try it with a RPI. But the tutorial left me some questions:
* how to prepare the rom file for a Xiaofang 1s with 64MB or 128 MB. I`m not sure, which one I have. 
* where to get the rom.layout file for flashrom
* how to set the other parameters of flashrom when I use a RPI instead of a ch341

**Comment by goermezer** [comment]
Ok, no need for help. I could do it on my own. I prepared the compiled 64MB bootloader for Xiaofang 1s with dd from the tutorial:
```
cp -r xiaofang1s_64mb_v2.bin full-xiaofang1s_64mb_v2.bin
dd if=/dev/zero of=full-xiaofang1s_64mb_v2.bin bs=1 count=1 seek=16777215
```
The file [rom.layout](https://github.com/Dafang-Hacks/spiflasher/blob/master/rom.layout) was attached to this tutorial. I have overseen it before. 

Parameters for flashrom look a bit different for SPI interface. I followed this tutorial https://tomvanveen.eu/flashing-bios-chip-raspberry-pi/ and came out with:

`flashrom -p linux_spi:dev=/dev/spidev0.0,spispeed=512 -n -V -c GD25Q128C --layout rom.layout --image boot -w full-xiaofang1s_64mb_v2.bin`

It took ~5-10 minutes until the boot rom was read, erased and written to flash. After soldering the chip back to the camera, it booted :-)

**Comment by hoxsiew** [comment]
@goermezer, good to hear that you got it back to booting.  I managed to get a shell on mine first, then just dd'd /dev/mtdblock[0-5] to a blank SD card.  Basically all those concatenated together make up the image on the flash chip.  When I fixed mine, I cat'd the originals all together and used flashrom to flash the entire block.  I suppose I could have done it in steps, or just the messed up section.

/proc/mtd shows how the flash chip is partitioned out:

> /tmp # cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00040000 00008000 "boot"
mtd1: 00280000 00008000 "kernel"
mtd2: 006f0000 00008000 "rootfs"
mtd3: 00630000 00008000 "data"
mtd4: 00010000 00008000 "config"
mtd5: 00010000 00008000 "factory"

/proc/partitions shows the size in KiB:
>/tmp # cat /proc/partitions 
major minor  #blocks  name

  31        0        256 mtdblock0
  31        1       2560 mtdblock1
  31        2       7104 mtdblock2
  31        3       6336 mtdblock3
  31        4         64 mtdblock4
  31        5         64 mtdblock5

**Comment by hoxsiew** [comment]
Well, I seem to have hard-bricked my device.  I was flashing a root-fs image from the SDcard to the flash memory using u-boot and it went along nicely until it read 84% and just locked up.  Lights went out, won't reset, won't do anything after power cable cycling; just plain dead.

Fun while it lasted.  Looks like I'll be going back to my hacking on my Grain-Media devices.

**Comment by pcmester** [comment]
@hoxsiew How did you get to root shell? I started /bin/ash, but everything is read-only, and i can not do anything as root, because i get this error:
`sh: can't access tty; job control turned off`

**Comment by hoxsiew** [comment]
@pcmester, It's been awhile, and I bricked the device so It's in a box of junk somewhere right now, but I recall that I just changed the last line of /init (which is a script) to launch /bin/ash instead of whatever it launched at that point.  I did this on a filesystem dump of the ramdisk image on my desktop machine, then created a new ramdisk image and flashed it to the memory chip with RPi zero eeprom programmer.

**Comment by laviniug** [comment]
Hi, is there anyone working on a simple way to root this without opening it up or should I simply sell it? Thanks

**Comment by pcmester** [comment]
I tired to resolder the chip, but I damaged it, and removed some resistors too. Now it connects to the PC as USB boot device. Could somebody help me with that?

**Comment by balonchiks** [comment]
hi, does the hack work for the same camera, but with the [black back](http://en.miui.com/thread-909532-1-1.html), that was released in 2017?

**Comment by jnsw** [comment]
is this dead?

**Comment by ciB89** [comment]
Can someone confirm / comment on using dafang hacks with SXJ02ZM? Openfang (the project that supposedly works with this camera) is dead and only works partially with the camera. I would like to switch to dafang hacks.

**Comment by hoxsiew** [comment]
> Can someone confirm / comment on using dafang hacks with SXJ02ZM? Openfang (the project that supposedly works with this camera) is dead and only works partially with the camera. I would like to switch to dafang hacks.

The image files and bootloaders referenced in the procedures for this project never worked with my camera.  I tried hacking some stuff together myself from files included and was able to root the device, but didn't get very far before I bricked the mainboard.  Maybe you'll get lucky and it will work with yours, but this project is unofficially dead and I wouldn't expect much support.
